File: mouvements/outils.py
import numpy as np
import time
import math


def calculer_distance(point_a, point_b):
    return math.sqrt((point_a.x - point_b.x) ** 2 + (point_a.y - point_b.y) ** 2)

def calculer_angle(a, b, c):
    """
    Calcule l'angle formé par trois points.

    a = premier point
    b = point central (articulation)
    c = troisième point
    """

    a = np.array([a.x, a.y])
    b = np.array([b.x, b.y])
    c = np.array([c.x, c.y])

    radians = (
        np.arctan2(c[1] - b[1], c[0] - b[0])
        -
        np.arctan2(a[1] - b[1], a[0] - b[0])
    )

    angle = abs(radians * 180 / np.pi)

    if angle > 180:
        angle = 360 - angle

    return angle

class HoldPosition:

    def __init__(self, position, duree):

        self.position = position
        self.duree = duree
        self.debut = None
        self.termine = False


    def update(self, corps):

        # La position n'est pas tenue
        if not self.position(corps):

            self.debut = None
            self.termine = False

            return 0, False


        # Début du maintien
        if self.debut is None:

            self.debut = time.time()


        # Temps écoulé
        temps_ecoule = time.time() - self.debut


        # Progression entre 0 et 100
        progression = (temps_ecoule / self.duree) * 100
        progression = min(progression,100)


        # Maintien terminé
        if progression >= 100:

            if not self.termine:

                self.termine = True

                return 100, True

            return 100, False


        return progression, False

File: web/static/js/outils.js
// Jumeau de mouvements/outils.py.

export function calculer_distance(point_a, point_b) {
  return Math.sqrt((point_a.x - point_b.x) ** 2 + (point_a.y - point_b.y) ** 2);
}

export function calculer_angle(a, b, c) {
  const radians =
    Math.atan2(c.y - b.y, c.x - b.x) - Math.atan2(a.y - b.y, a.x - b.x);
  let angle = Math.abs((radians * 180) / Math.PI);
  if (angle > 180) {
    angle = 360 - angle;
  }
  return angle;
}

export class HoldPosition {
  // L'horloge est injectable pour que la classe reste testable hors
  // navigateur : le harnais de comparaison ne peut pas attendre 1,5 s reelle.
  constructor(position, duree, horloge = () => performance.now() / 1000) {
    this.position = position;
    this.duree = duree;
    this.horloge = horloge;
    this.debut = null;
    this.termine = false;
  }

  update(corps) {
    if (!this.position(corps)) {
      this.debut = null;
      this.termine = false;
      return [0, false];
    }

    if (this.debut === null) {
      this.debut = this.horloge();
    }

    const temps_ecoule = this.horloge() - this.debut;
    const progression = Math.min((temps_ecoule / this.duree) * 100, 100);

    if (progression >= 100) {
      if (!this.termine) {
        this.termine = true;
        return [100, true];
      }
      return [100, false];
    }

    return [progression, false];
  }
}
